### PR TITLE
Linear and Conv2d links support float16 type, and example is added

### PR DESCRIPTION
**Pull request by okuta**
Please merge #1295  before this PR.
I implemented float16 examples (`alex_fp16` and `googlenetbn_fp16`).

### CODE REVIEW

File: chainer/link.py [hunk 1]
```diff
@@ -130,15 +133,21 @@ def add_param(self, name, shape, dtype=numpy.float32):
                 name will be removed.
             shape (int or tuple of ints): Shape of the parameter array.
             dtype: Data type of the parameter array.
+            initializer(chainer.initializer.Initializer): If not ``None``,
+                the data is initialized by initializer. In that case,
+                ``dtype`` argument is ignored.
```
Comment (rezoo) [inline chainer/link.py]: If it is not ``None``, the data is initialized with the given initializer.
Note that in this case ``dtype`` argument is ignored.

File: chainer/link.py [hunk 2]
```diff
@@ -130,15 +133,21 @@ def add_param(self, name, shape, dtype=numpy.float32):
...
             raise AttributeError(
                 'cannot register a new parameter %s: attribute exists'
                 % name)
-        data = self.xp.full(shape, numpy.nan, dtype=dtype)
-        grad = data.copy()
+        if initializer is None:
+            data = self.xp.full(shape, numpy.nan, dtype=dtype)
+        else:
+            data = initializers.generate_array(initializer, shape, self.xp)
+        grad = self.xp.full_like(data, numpy.nan)
```
Comment (rezoo) [inline chainer/link.py]: It is well-known that filling the gradients with ``np.nan`` is important to avoid bugs, however, I also think that is slightly breaks backward compatibility when the users use the wrong implementation.
Is this modification ok for the chainer team?

File: chainer/links/connection/convolution_2d.py [hunk 3]
```diff
@@ -55,8 +55,10 @@ def __init__(self, in_channels, out_channels, ksize, stride=1, pad=0,
         self.pad = _pair(pad)
         self.use_cudnn = use_cudnn
         self.out_channels = out_channels
-        self.initialW = initialW
-        self.wscale = wscale
+        # For backward compatibility, the scale of weights is proportional to
+        # the square root of wscale.
+        self.W_initializer = initializers._get_initializer(
+            initialW, scale=math.sqrt(wscale))
```
Comment (rezoo) [inline chainer/links/connection/convolution_2d.py]: The same as above; for the users who use the member variables ``initialW`` and ``wscale``, it breaks backward compatibility. I think this PR should be merged at the next major release.

File: chainer/links/connection/inceptionbn.py [hunk 4]
```diff
@@ -33,6 +35,8 @@ class InceptionBN(link.Chain):
             ``cupy.ndarray`` and edits its value.
             It is used for initialization of the convolution matrix weights.
             Maybe be ``None`` to use default initialization.
+        dtype (numpy.dtype): Type to use in
+            ``~batch_normalization.BatchNormalization``.
```
Comment (rezoo) [inline chainer/links/connection/inceptionbn.py]: Why ``dtype`` is only used in ``BatchNormalization`` layers, and not used in convolutional layers?

File: chainer/links/connection/linear.py [hunk 5]
```diff
@@ -47,9 +47,11 @@ class Linear(link.Link):
     def __init__(self, in_size, out_size, wscale=1, bias=0, nobias=False,
                  initialW=None, initial_bias=None):
         super(Linear, self).__init__()
-        self.initialW = initialW
-        self.wscale = wscale
```
Comment (rezoo) [inline chainer/links/connection/linear.py]: Same as above.

File: examples/imagenet/alex.py [hunk 6]
```diff
@@ -37,3 +40,30 @@ def __call__(self, x, t):
         loss = F.softmax_cross_entropy(h, t)
         chainer.report({'loss': loss, 'accuracy': F.accuracy(h, t)}, self)
         return loss
+
+
+class Alex_fp16(Alex):
```
Comment (rezoo) [inline examples/imagenet/alex.py]: How about using AlexFp16 instead of underscore?

File: examples/imagenet/alex.py [hunk 7]
```diff
@@ -37,3 +40,30 @@ def __call__(self, x, t):
...
+            conv4=L.Convolution2D(384, 384,  3, pad=1, initialW=W, bias=bias),
+            conv5=L.Convolution2D(384, 256,  3, pad=1, initialW=W, bias=bias),
+            fc6=L.Linear(9216, 4096, initialW=W, bias=bias),
+            fc7=L.Linear(4096, 4096, initialW=W, bias=bias),
+            fc8=L.Linear(4096, 1000, initialW=W, bias=bias),
+        )
+        self.train = True
+
+    def __call__(self, x, t):
+        return Alex.__call__(self, F.cast(x, self.dtype), t)
```
Comment (rezoo) [inline examples/imagenet/alex.py]: How about adding the class variable ``dtype`` to ``Alex``, and simplify ``AlexFp16`` by modifying ``dtype`` like that?

```python
class Alex(chainer.Chain):

    """Single-GPU AlexNet without partition toward the channel axis."""

    dtype = np.float32
    insize = 227

    def __init__(self):
        W = initializers.HeNormal(1 / np.sqrt(2), self.dtype)
        bias = initializers.Zero(self.dtype)
        super(Alex, self).__init__(
            conv1=L.Convolution2D(3,  96, 11, stride=4),
            conv2=L.Convolution2D(96, 256,  5, pad=2),
            conv3=L.Convolution2D(256, 384,  3, pad=1),
            conv4=L.Convolution2D(384, 384,  3, pad=1),
            conv5=L.Convolution2D(384, 256,  3, pad=1),
            fc6=L.Linear(9216, 4096),
            fc7=L.Linear(4096, 4096),
            fc8=L.Linear(4096, 1000),
        )
        self.train = True

    def __call__(self, x, t):
        x = F.cast(x, self.dtype)
        h = F.max_pooling_2d(F.relu(
            F.local_response_normalization(self.conv1(x))), 3, stride=2)
        h = F.max_pooling_2d(F.relu(
            F.local_response_normalization(self.conv2(h))), 3, stride=2)
        h = F.relu(self.conv3(h))
        h = F.relu(self.conv4(h))
        h = F.max_pooling_2d(F.relu(self.conv5(h)), 3, stride=2)
        h = F.dropout(F.relu(self.fc6(h)), train=self.train)
        h = F.dropout(F.relu(self.fc7(h)), train=self.train)
        h = self.fc8(h)

        loss = F.softmax_cross_entropy(h, t)
        chainer.report({'loss': loss, 'accuracy': F.accuracy(h, t)}, self)
        return loss


class AlexFp16(Alex):
    dtype = np.float16
```

File: examples/imagenet/googlenetbn.py [hunk 8]
```diff
@@ -105,3 +108,61 @@ def __call__(self, x, t):
             'accuracy': accuracy,
         }, self)
         return loss
+
+
+class GoogLeNetBN_fp16(GoogLeNetBN):
```
Comment (rezoo) [inline examples/imagenet/googlenetbn.py]: How about using ``GoogleLeNetBNFp16``?

File: examples/imagenet/googlenetbn.py [hunk 9]
```diff
@@ -105,3 +108,61 @@ def __call__(self, x, t):
...
+            convb=L.Convolution2D(576, 128, 1, initialW=W, nobias=True),
+            normb=L.BatchNormalization(128, dtype=dtype),
+            linb=L.Linear(2048, 1024, initialW=W, nobias=True),
+            normb2=L.BatchNormalization(1024, dtype=dtype),
+            outb=L.Linear(1024, 1000, initialW=W, bias=bias),
+        )
+        self._train = True
+
+    def __call__(self, x, t):
+        return GoogLeNetBN.__call__(self, F.cast(x, self.dtype), t)
```
Comment (rezoo) [inline examples/imagenet/googlenetbn.py]: Same as above ``AlexFp16``.

**Comment by rezoo** [comment]
Thats all. Please clarify the above questions.

File: chainer/links/connection/inceptionbn.py (hunk 4 above)
Comment (okuta) [inline chainer/links/connection/inceptionbn.py]: Convolutional layer use `conv_init`.

File: examples/imagenet/alex.py (hunk 7 above)
Comment (okuta) [inline examples/imagenet/alex.py]: I think that current code is easy to understand. Users may want simple example.
How do you think?

**Comment by okuta** [comment]
Thank you for your reviewing. I fixed.

File: chainer/link.py (hunk 2 above)
Comment (beam2d) [inline chainer/link.py]: Why does it break compatibility here? It looks `grad` is initialized by NaN in the old code, too (or I'm missing something?).

Comment (rezoo) [inline chainer/link.py]: Ah sorry I misunderstand. Please just ignore the above comment.

Comment (beam2d) [inline chainer/link.py]: OK, no problem.

File: chainer/links/connection/inceptionbn.py (hunk 4 above)
Comment (rezoo) [inline chainer/links/connection/inceptionbn.py]: I see.

File: examples/imagenet/alex.py (hunk 7 above)
Comment (rezoo) [inline examples/imagenet/alex.py]: Users may want simple example

I think the code I have proposed is easy to understand and more simpler, but I do not want to discuss about its superiority or inferiority.

**Comment by beam2d** [comment]
LGTM
